---
title: "PATH302 - Breast Cancer Data"
author: "Mik Black"
date: "24 August 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this lab we will be using the R computing environment to perform a basic analysis of
gene expression data from a breast cancer data set.

The data we will be using comes from the following publication:

Miller, L. D., Smeds, J., George, J., Vega, V. B., Vergara, L., Ploner, A., et al. (2005). An expression signature for p53 status in human breast cancer predicts mutation status, transcriptional effects, and patient survival. _Proceedings of the National Academy of Sciences of the United States of America_, 102(38), 13550â€“13555. 

It comprises microarray and clinical data generated from tumour samples obtained from 251 female breast cancer patients from Uppsala County, Sweden, from January 1, 1987 to December 31, 1989.  While this is a relatively old data set, it still reflects the type of data generated in clinical genomics studies today (i.e., sequencing-based transcriptomic data from tumours).

## Setup 

Click on the Windows Start Menu in the bottom left of the Student Desktop, and open the RStudio application.

### Installation

In order to access the data and commands needed for this lab, we first need to install some add-on packages.
Run the commands below in R to do this (you can just cut and paste these commands into the R console - if you're not sure what this means, just let me know):

```{r, eval=F}
dir.create('Rlibs')
newPath = c(.libPaths(), paste0(getwd(), "/Rlibs"))
.libPaths(newPath)

install.packages('BiocInstaller', repos='http://www.bioconductor.org/packages/3.4/bioc', lib='Rlibs')
library(BiocInstaller, lib.loc="Rlibs")
biocLite("Biobase", lib="Rlibs")
library("Biobase", lib.loc="Rlibs")
```

### Load packages

The following commands load packages that we will use in our analysis.

```{r, warning=FALSE, message=FALSE}
library(ggplot2)
library(dplyr)
library(gplots)
library(survival)
library(httr)
```

### Load data

Unfortunately, due to the way in which the student desktop is set up, the package that contains that data,
`breastCancerUPP` can't easily be installed.  I have instead created a data file containing the required 
information, that can be loaded via the following command:

```{r, echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
load(url("https://github.com/mikblack/PATH302/raw/master/uppsalaCohort.RData"))
```

```{r, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
library(Biobase)
load("uppsalaCohort.RData")
```

This loads a data object called `upp` which holds LOTS of relevant information about the Uppsala breast cancer cohort.  We are interested in data about: gene expression, microarray probe annotation (so we can understand what gene each microarray probe relates to) and clinical information about the patients.  It also loads an object called 
`PAM50Preds` which we'll use later.

### Extract cohort data

Gene expression data:

```{r}
uppExp   = exprs(upp)
```

View data using RStudio built-in viewer

```{r, eval=FALSE}
View(uppExp)
```

Gene annotation data

```{r}
uppAnnot = featureData(upp)@data
```

```{r, eval=FALSE}
View(uppAnnot)
```

Clinical data:

```{r}
uppClin  = phenoData(upp)@data
```

```{r, eval=FALSE}
View(uppClin)
```

### Clinical data

From looking at the clinical data (using the  `View()` command), it is clear that some of the variables
contain no useful information.  Let's select a subset of variables:

```{r}
uppClinSmall = uppClin[,c("size", "age", "er", "grade", "pgr", 
                          "node", "t.rfs", "e.rfs","treatment")]
```

Now we have a smaller data object to work will.  Take a look:

```{r, eval=FALSE}
View(uppClinSmall)
```

Note that the "recurrence free survival time", `rfs.t` variable is given in "days", but "years" is
easier to interpret.  Convert RFS time to years (from 'days'):

```{r}
uppClinSmall$t.rfs = uppClinSmall$t.rfs / 365
```

Attach data so that we can directly access the clinical variables by name

```{r}
attach(uppClinSmall)
```

For example:

```{r}
grade
```

The clinical variables are:

```{r, echo=FALSE}
desc = c("Tumour Size (cm)",
         "Patient age at diagnosis (years)",
         "Estrogen receptor status (0=negative, 1=positive)",
         "Tumour Grade (1, 2, or 3)",
         "Progesterone receptor status (0=negative, 1=positive)",
         "Lymph node status (0=negative, 1=positive)",
         "Recurrence free survival time (years)",
         "Recurrence free survival event (0=censored, 1=recurrence)",
         "Treatment (0=none, 2=endocrine treatment or chemotherapy")
datTab = data.frame(Variable = names(uppClinSmall),
                    Description = desc)
knitr::kable(datTab)
```


#### Plots of clinical data

Let's make some plots: we are using the `ggplot2` framework for plotting.  The first part of each plot specification (the `ggplot` part) provides information at the data object to be used (`uppClinSmall`) and the variable(s) of interest (`size`).  The next piece (the `geom_` part) tells R what sort of plot we want to make - in this case it is a histogram (`geom_histogram()`).  The details inside the `geom_histogram` function specify the bar colours (outline and fill), and the `xlab`, `ylab` and `ggtitle` commands specify the axis labels and plot title:

```{r, warning=FALSE, message=FALSE}
ggplot(data=uppClinSmall, aes(x=size)) + 
  geom_histogram(fill="white", colour="black") +
  xlab("Tumour Size") +
  ylab("Frequency") + 
  ggtitle("Uppsala cohort: tumour size")
```

__Challenge 1:__ _modify the code above to produce the following graph:_

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(data=uppClinSmall, aes(x=age)) + 
  geom_histogram(fill="white", colour="black") +
  xlab("Age") +
  ylab("Frequency") + 
  ggtitle("Uppsala cohort: patient age")
```

There are __LOTS__ of plots we can make.  Here's a boxplot showing the distribution of tumour size across 
tumour grade:

```{r}
ggplot(data=uppClinSmall, aes(x=as.factor(grade),y=size)) + 
  geom_boxplot()
```

Get rid of the annoying large `size` value using the `ylim` command to change the limits on the y-axis, 
and the `NA` values in `grade` using the `subset` command:

```{r, warning=FALSE, message=FALSE}
ggplot(data=subset(uppClinSmall, !is.na(grade)), aes(x=as.factor(grade),y=size)) + 
  geom_boxplot() +
  ylim(0,7) +
  ggtitle("Uppsala cohort: tumour size versus grade")
```

__Challenge 2:__ _Modify the code above to generate the following plot:_

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(data=subset(uppClinSmall, !is.na(er)), aes(x=as.factor(er),y=size)) + 
  geom_boxplot() + 
  ylim(0,7) +
  ggtitle("Uppsala cohort: tumour size versus age")
```

What do you conclude about the relationship between ER status and tumour size?

#### Tables of clinical data

We can also use R to generate tables relating to the clinical variables.  For example, ER status:

```{r}
table(er)
```

We can also calculate proportions using the `prop.table` command:

```{r}
table(grade)
prop.table( table(grade) )
```

Adding a second variable to the `table` command creates a two-way contingency table:

```{r}
table(er, grade)
```

The `prop.table` command then provides multiple options:

```{r}
## Overall proportions:
prop.table(table(er, grade))

## Proportions by row (rows sum to 1)
prop.table(table(er, grade),1)

## Proportions by columns (columns sum to 1)
prop.table(table(er, grade),2)
```

If we want to test whether two variables are significantly associated, Fisher's Exact Test can be applied to the contingency table:

```{r}
fisher.test(table(er, grade))
```

Alternatively the Chi-squared test (which is an approximation) can be used:

```{r}
chisq.test(table(er, grade))
```

__Challenge 3:__ _Generate a table of ER status versus Lymph Node Status, and determine whether they are significantly associated with each other._

### Molecular subtyping

We can use the microarray data to generate a _molecular subtype_ for each tumour.  To do this we use the 
`molecular.subtyping` command from the `genefu` package.  There are a number of subtyping methods available - we are using the popular "PAM50" approach, which uses the expression patterns of collection of 50 genes to put
each tumour into one of five subtypes: Basal-like, Her2/Neu, Luminal A, Luminal B and Normal-like.  The command parameters specify the subtyping model to be used (`pam50`), the gene expression data (`uppExp` - note that the matrix needs to be transposed (rotated 90 degrees), hence the `t()` function), and annotation information linking probes to genes (`uppAnnot`).  You can use the R help facility (via `?molecular.subtyping`) to find out what the `do.mapping` parameter does, if you are interested.  :)

__YOU DON'T NEED TO RUN THE NEXT BLOCK OF CODE.__

__the `genefu` package that performs the molecular subtyping is annoying to install, because it has a large number of package dependencies.  As a results, I have peformed this for you, using the code below, and the object 
`PAM50Preds` was loaded with the Uppsala cohort data above.__

```{r, eval=FALSE}
library(genefu)
PAM50Preds = molecular.subtyping(sbt.model = "pam50", data=t(uppExp),
                                annot=uppAnnot, do.mapping=TRUE)
```

This produces the following subtype information:

```{r}
table(PAM50Preds$subtype)
```

We can add this information to the clinical data:

```{r, warning=FALSE, message=FALSE}
uppClinSmall$subtype = PAM50Preds$subtype
attach(uppClinSmall)
```

Compare Subtype to ER status:

```{r}
table(subtype, er)

table(subtype, er) %>% 
  prop.table(., 1) %>% 
  round(., 3)
```

Note - the use of the "pipe" operator, `%>%` illustrates a fancy (and more readable) way of writing

```{r}
round( prop.table( table(subtype, er), 1) ,3)
```

Which approach do you prefer?

__Challenge 4:__ _Modify the code above to investigate the relationship between Grade and Subtype. Which subtype has the highest proportion of Grade 3 tumours?  Is the association between Grade and Subtype statistically significant?_

```{r, echo=FALSE, eval=FALSE}
table(subtype, grade)
table(subtype, grade) %>% 
  prop.table(., 1) %>% 
  round(., 3)
```

### Survival analysis

We can use the `survival` package (installed as part of base R, and loaded above) to examine the 
survival characteristics of patients in the Uppsala breast cancer cohort.  Here we have "Recurrence Free Survival" data available, so a "survival event" relates to cancer relapse: either local of distant metastasis.  The "survival time" indicates either the number of years after initial treatment (surgery) that cancer recurrence
was detected, or (if there has been no recurrence), the number of years after surgery that the patient was last observed.  Patients who were disease-free when they were last seen are considered "censored", because we don't know their current status.

```{r}
plot( survfit(Surv(t.rfs, e.rfs) ~1 ), 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")
```

```{r}
plot( survfit(Surv(t.rfs, e.rfs) ~ er ), col=1:2, 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")
legend('bottomleft', c("ER-", "ER+"), fill=1:2)
```

__Challenge 5:__ _Modify the code above to generate a survival plot for tumour grade.  Does relationship between grade and recurrence free survival match what you would expect?_

```{r, echo=FALSE, eval=FALSE}
plot( survfit(Surv(t.rfs, e.rfs) ~ grade ), col=1:3, 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")
legend('bottomleft', c("G1", "G2", "G3"), fill=1:3)
```

The following is a survival plot for tumour subtype.

```{r]}
plot( survfit(Surv(t.rfs, e.rfs) ~ subtype ), col=1:5, 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")
groups <- names(table(subtype))
legend('bottomleft', groups, fill=1:5)
```

We can test to see whether the relationship between subtype and survival is significant using the `survdiff`
command:

```{r}
survdiff(Surv(t.rfs, e.rfs) ~ subtype)
```

We can see that the result is highly significant, so we conclude that the different tumour subtypes are 
associated with differences in recurrence free survival rates.

__Challenge 5:__ _Modify the code above to test for an association between grade and survival.  What do you conclude?_

```{r, echo=FALSE, eval=FALSE}
survdiff(Surv(t.rfs, e.rfs) ~ grade)
```

### Gene expression data

Up to this point, we have only been looking at clinical data for the Uppsala cohort (although we did use the
gene expression information to determine the tumour subtypes).

Here we are going to have a look at a subset of the gene expression data for each tumour, and examine some
associations with clinical information.  We'll start by looking at gene expression for ESR1, a gene that
encodes an estrogen receptor.

Let's find the microarray probe(s) for ESR1:

```{r}
esr1Probes = uppAnnot$probe[ na.omit(uppAnnot$Gene.symbol == 'ESR1') ]
esr1Probes
```

```{r, eval=FALSE, echo=FALSE}
par(mfrow=c(2,5))
for(i in esr1Probes){
  x = uppExp[match(i, rownames(uppExp)),]
  boxplot(x ~ er)
}
```

There are multiple probes, but the best one to use is the first one: `esr1Probes[1]` (trust me).  

We need to then figure out what row of the gene expression data this relates to:

```{r}
match("205221_at", rownames(uppExp))
```

Then we can extract the data from this row:

```{r}
esr1Dat = uppExp[match("205221_at", rownames(uppExp)), ]
```

To make things easy, lets add this variable to our clinical data:

```{r, warning=FALSE, message=FALSE}
uppClinSmall$esr1Dat = esr1Dat
attach(uppClinSmall)
```

Now we can make a boxplot to examine ESR1 expression across ER- (0) and ER+ (1) tumours:

```{r}
ggplot(data=subset(uppClinSmall, !is.na(er)), aes(x=as.factor(er), y=esr1Dat)) + geom_boxplot()
```

__Challenge 6:__ _Explain what the boxplot above shows.  Is this what you would expect?_

### Proliferation genes

The genes below are involved in cellular proliferation, and proliferation plays an important role in the growth and development of tumours.

```{r, echo=FALSE}
#prolifGenes = c("OSTC","MCM6","RPA3","MCM7","PCNA","XRCC6","KPNA2","ANLN","RNASEH2A",
#                "PBK","GMNN","RRM1","CDC45","MAD2L1","RAN","DUT","RRM2","CDK7",
#                "MLH3","SMC4","SMC3","POLD2","POLE2","BCCIP","GINS2","TREX1",
#                "BUB3","FEN1","DBF4B","MOB4","CCNE1","RPA1","POLE3","RFC4","MCM3",
#                "CHEK1","CCND1","CDC37")
prolifGenes = c("MAD2L1", "RRM2", "ANLN", "MCM6", "PBK", "GINS2", "KPNA2", "PCNA")
knitr::kable(data.frame(Gene=prolifGenes, 
                        Probe=uppAnnot$probe[match(prolifGenes, uppAnnot$Gene.symbol)]))
```

We can use the `match` command to find which rows of the gene expression data set correspond to these proliferation genes.  Note that using `match` only finds the first match to the gene name - if there are multiple probes for a gen, then we will only retrieve the first one.  This is not the best way to operate, but it will suffice for what we are doing here (in practice, we often merge probe data together when there a multiple probes representing a single gene).

```{r}
prolifRows = na.omit(match(prolifGenes, uppAnnot$Gene.symbol))
```

Here are the rows of the gene expression data set that relate to our proliferation genes:

```{r}
prolifRows
```

Extract the data for these rows:

```{r}
prolifDat = uppExp[prolifRows, ]
```

The following code is a little ugly, but it does two important things:

1. it transforms each gene to have a mean of zero and a standard deviation of 1 (i.e., it puts all of the genes on the  same scale, which helps with data visualisation).

2. it sets any transformed value above 3 (or less than -3) to 3 (-3) so that there are no extreme values (this makes sure that the range of our visualisation isn't too great.)

```{r}
prolifDatScale = t(scale(t(prolifDat)))
prolifDatScale[prolifDatScale < -3] = -3
prolifDatScale[prolifDatScale >  3] =  3
```

After transforming the data, we can visualise the gene expression information for the proliferation genes using a _heatmap_.  Although R has a built-in heatmap command (`heatmap`), the `gplots` package has a greatly improved
function (imaginatively titled `heatmap.2`) that we will use:

```{r}
heatmap.2(prolifDatScale, trace='none', scale='none', col='bluered', 
          labRow = uppAnnot$Gene.symbol[prolifRows])
```

__Challenge 7:__ _Explain what is being shown in the heatmap._

Since all of the proliferation genes are basically behaving in the same way, we can summarise their
activity by taking their mean for each sample.  This type of summary is called a "centroid".  We use the `colMeans` function to take the mean of each column (i.e., per-patient mean across the proliferation genes):

```{r}
prolifMean = colMeans(prolifDat)
```

This generates a single value for each tumour, which can be thought of as an indication of the level of 
proliferative activity.  To see how this summary relates to the original data, we can add the information above the heatmap, and then order the columns of the heatmap according to the centroid values (instead of using the
default clustering provided by the `heatmap.2` command):

First, get the order of the centroid values, and create colors relating to those values:

```{r}
ord = order(prolifMean)
prolifCol = bluered(length(prolifMean))[rank(prolifMean)]
```

Next, use the `heatmap.2` command again, but this time tell it not to do the clustering for the columns 
(`Colv=FALSE`), and instead to use the `ord` variable to determine the ordering of the columns and the colour object (`prolifCol`):

```{r}
heatmap.2(prolifDatScale[,ord], trace='none', scale='none', col='bluered', 
          labRow = uppAnnot$Gene.symbol[prolifRows], Colv=FALSE,
          ColSideColors=prolifCol[ord])
```

Now we can clearly see that low values of the centroid (blue) correspond to uniformly low expression levels for the proliferation genes, and vice versa (high centroid value (red), corresponds to high gene expression).  This suggests that our proliferation centroid can be used as an effective summary of the activity of those proliferation genes.  How can we use this information?

Having a one-dimensional summary like a centroid (i.e., one value per tumour) allows us to investigate the relationship between proliferation and other variables we have available, for example, our clinical variables. Lets add the proliferation centroid to our clinical data:

```{r, warning=FALSE, message=FALSE}
uppClinSmall$prolifMean = prolifMean
attach(uppClinSmall)
```


Here is a boxplot showing the distribution of the proliferation centroid across tumour grade.

```{r}
ggplot(data=subset(uppClinSmall, !is.na(grade)), aes(x=as.factor(grade), y=prolifMean)) + geom_boxplot()
```

__Challenge 8:__ _Modify the above code to investigate the relationship between proliferation and tumour subtype.  What does the plot tell you?_

```{r, echo=FALSE, eval=FALSE}
ggplot(data=uppClinSmall, aes(x=as.factor(subtype), y=prolifMean)) + geom_boxplot()
```

Lastly, we can use the centroid to define groups of tumours with particular characteristics.  Here we will 
split the tumours into two groups, one exhibiting "low" proliferation" (values below centroid median), and one exhibiting "high" proliferation" (values above centroid median).

```{r}
prolifHilo = ifelse(prolifMean > median(prolifMean), "HighProilif", "LowProlif")
```

This puts half the tumours in the "low" group, and half in the "high" group:

```{r}
table(prolifHilo)
```

Now we can investigate the relationship with other variables, for example, grade:

```{r}
table(grade, prolifHilo)
```

We can also investigate the relationship between proliferation and patient survival:

```{r}
plot( survfit(Surv(t.rfs, e.rfs) ~ prolifHilo ), col=1:2, 
      xlab="Time (years)", ylab = "Proportion recurrence free")
groups <- names(table(prolifHilo))
legend('bottomleft', groups, fill=1:2)
```

__Challenge 9:__ _What does the plot above tell you about the relationship between proliferative activity in the tumour, and recurrence free survival?_

### IN CASE OF INSTALL DISASTER

If we have a problem with installing packages at the start of the session, this command:

```{r, eval=FALSE}
load(url("https://github.com/mikblack/PATH302/raw/master/uppsalaObjects.RData"))
```

will load a file contains the `uppExp`, `uppAnnot` and `uppClin` objects, which will remove the need to have 
the Bioconductor packages installed.
