---
title: "PATH302 - Breast Cancer Data"
author: "Mik Black"
date: "24 August 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this lab we will be using the R computing environment to perfom a basic analysis of
gene expression data from a breast cancer data set.

The data we will be using comes from the following publication:

Miller, L. D., Smeds, J., George, J., Vega, V. B., Vergara, L., Ploner, A., et al. (2005). An expression signature for p53 status in human breast cancer predicts mutation status, transcriptional effects, and patient survival. Proceedings of the National Academy of Sciences of the United States of America, 102(38), 13550â€“13555. 

It comprises microarray and clinical data generated from tumour samples obtained from 251 female breast cancer patients from Uppsala County, Sweden, from January 1, 1987 to December 31, 1989.  While this is a relatively old data set, it still reflects the type of data generated in clinical genomics studies today (i.e., sequencing-based transcriptomic data from tumours).

## Setup 

Click on the Windows Start Menu in the bottom left of the Student Desktop, and open the RStudio application.

### Installation

In order to access the data and commands needed for this lab, we first need to install some add-on packages.
Run the commands below in R to do this (you can just cut and paste these commands into the R console - if you're not sure what this means, I'll show you on projector):

```{r, eval=F}
dir.create('Rlibs')
newPath = c(.libPaths(), paste0(getwd(), "Rlibs"))
.libPaths(newPath)

install.packages('BiocInstaller', repos='http://www.bioconductor.org/packages/3.4/bioc', lib='Rlibs')

install.packages('dplyr', lib='Rlibs')
install.packages('ggplot2', lib='Rlibs')
install.packages('gplots', lib='Rlibs')
install.packages('dplyr', lib='Rlibs')

biocLite("genefu", lib="Rlibs")
```

### Load packages

```{r, warning=FALSE, message=FALSE}
library(genefu)
library(ggplot2)
library(dplyr)
library(gplots)
```

### Load data

Unfortunately, due to the way in which the student desktop is set up, the package that conatins that data,
`breastCancerUPP` can't eaily be installed.  I have instead created a data file containing the reqired 
information, that can be loaded via the following command:

```{r, echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
library(httr)
load(url("https://github.com/mikblack/PATH302/raw/master/uppsalaCohort.RData"))
```

```{r, echo=FALSE, eval=TRUE}
load("uppsalaCohort.RData")
```

Now that the data file has been loaded, we can access the Uppsala cohort data:

```{r}
data(upp)
```

This loads a data object called `upp` which holds LOTS of relevant information.  We are interested in data about: gene expression, microarray probe annotation (so we can understand what gene each microarray probe relates to) and clinical information about the patients.

### Extract cohort data

Gene expression data:

```{r}
uppExp   = exprs(upp)
```

View data using RStudio built-in viewer

```{r, eval=FALSE}
View(uppExp)
```

Gene annotation data

```{r}
uppAnnot = featureData(upp)@data
```

```{r, eval=FALSE}
View(uppAnnot)
```

Clinical data:

```{r}
uppClin  = phenoData(upp)@data
```

```{r, eval=FALSE}
View(uppClin)
```

### Clinical data

From looking at the clinical data (using the  `View()` command), it is clear that some of the variables
contain no useful information.  Let's select a subset of variables:

```{r}
uppClinSmall = uppClin[,c("size", "age", "er", "grade", "pgr", 
                          "node", "t.rfs", "e.rfs","treatment")]
```

Now we have a smaller data object to work will.  Take a look:

```{r, eval=FALSE}
View(uppClinSmall)
```

Note that the "recurrence free survival time", `rfs.t` variable is given in "days", but "years" is
easier to interpret.  Convert RFS time to years (from 'days'):

```{r}
uppClinSmall$t.rfs = uppClinSmall$t.rfs / 365
```

Attach data so that we can directly access the clinical variables by name

```{r}
attach(uppClinSmall)
```

For example:

```{r}
grade
```

The clinical variables are:

```{r, echo=FALSE}
desc = c("Tumour Size (cm)",
         "Patient age at diagnosis (years)",
         "Estrogen receptor status (0=negative, 1=positive)",
         "Tumour Grade (1, 2, or 3)",
         "Progesterone receptor status (0=negative, 1=positive)",
         "Lymph node status (0=negative, 1=positive)",
         "Recurrence free survival time (years)",
         "Recurrence free survival event (0=censored, 1=recurrence)",
         "Treatment (0=none, 2=endocrine treatment or chemotherapy")
datTab = data.frame(Variable = names(uppClinSmall),
                    Description = desc)
knitr::kable(datTab)
```


#### Plots of clinical data

Let's make some plots: we are using the `ggplot2` framework for plotting.  The first part of each plot specification (the `ggplot` part) provides information at the data object to be used (`uppClinSmall`) and the variable(s) of interest (`size`).  The next piece (the `geom_` part) tells R what sort of plot we want to make - in this case it is a historgram (`geom_histogram()`).  The details inside the `geom_histogram` function specificy the bar colours (outline and fill), and the `xlab`, `ylab` and `ggtitle` commands specify the axis labels and plot title:

```{r, warning=FALSE, message=FALSE}
ggplot(data=uppClinSmall, aes(x=size)) + 
  geom_histogram(fill="white", colour="black") +
  xlab("Tumour Size") +
  ylab("Frequency") + 
  ggtitle("Uppsala cohort: tumour size")
```

__Challenge 1:__ _modify the code above to produce the following graph:_

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(data=uppClinSmall, aes(x=age)) + 
  geom_histogram(fill="white", colour="black") +
  xlab("Age") +
  ylab("Frequency") + 
  ggtitle("Uppsala cohort: patient age")
```

There are __LOTS__ of plots we can make.  Here's a boxplot:

```{r}
ggplot(data=uppClinSmall, aes(x=as.factor(grade),y=size)) + 
  geom_boxplot()
```

Get rid of the annoying large `size` value using the `ylim` command to change the limits on the y-axis, 
and the `NA` values in `grade` using the `subset` command:

```{r}
ggplot(data=subset(uppClinSmall, !is.na(grade)), aes(x=as.factor(grade),y=size)) + 
  geom_boxplot() +
  ylim(0,7) +
  ggtitle("Uppsala cohort: tumour size versus grade")
```

__Challenge 2:__ _Modify the code above to generate the following plot:_

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(data=subset(uppClinSmall, !is.na(er)), aes(x=as.factor(er),y=size)) + 
  geom_boxplot() + 
  ylim(0,7) +
  ggtitle("Uppsala cohort: tumour size versus age")
```

What do you conclude about the relationship between ER status and tumour size?

#### Tables of clinical data

We can also use R to generate tables realting to the clinical variables.  For example, ER status:

```{r}
table(er)
```

We can also calculate proportions using teh `prop.table` command:

```{r}
table(grade)
prop.table( table(grade) )
```

Adding a second variable to the `table` command creates a two-way contingency table:

```{r}
table(er, grade)
```

The `prop.table` command then provides multiple options:

```{r}
## Overall proportions:
prop.table(table(er, grade))

## Proportions by row (rows sum to 1)
prop.table(table(er, grade),1)

## Proportions by columns (columns sum to 1)
prop.table(table(er, grade),2)
```

If we want to test whether two variables are significantly associated, Fisher's Exact Test can be applied to teh contingency table:

```{r}
fisher.test(table(er, grade))
```

Alternatively the Chi-squared test (which is an approximation) can be used:

```{r}
chisq.test(table(er, grade))
```

__Challenge 3:__ _Generate a table of ER status versus Lymph Node Status, and determine whether they are significantly associated with each other._

### Molecular subtyping

We can use the microarray data to generate a _molecular subtype_ for each tumour.  To do this we use the 
`molecular.subtyping` command from the `genefu` package.  There are a number of subtyping methods available - we are using teh popular "PAM50" approach, which uses teh expression patterms of collection of 50 genes to put
each tumour into one of five subtypes: Basal-like, Her2/Neu, Luminal A, Luminal B and Normal-like.  The command parameters specify the subtyping model to be used (`pam50`), the gene expression data (`uppExp` - note that the matrix needs to be transposed (rotated 90 degrees), hence the `t()` function), and annotation information linking probes to genes (`uppAnnot`).  You can use the R help facility (via `?molecular.subtyping`) to find out what the `do.mapping` parameter does, if you are interested.  :)

```{r}
PAM50Preds = molecular.subtyping(sbt.model = "pam50", data=t(uppExp),
                                annot=uppAnnot, do.mapping=TRUE)
```

This produces the following subtype information:

```{r}
table(PAM50Preds$subtype)
```

We can add this information to the clinical data:

```{r, warning=FALSE, message=FALSE}
uppClin$subtype = PAM50Preds$subtype
attach(uppClin)
```

Compare Subtype to ER status:

```{r}
table(subtype, er)

table(subtype, er) %>% 
  prop.table(., 1) %>% 
  round(., 3)
```

Note - the use of the "pipe" operator, `%>%` illustrates a fancy (and more readable) way of writing

```{r}
round( prop.table( table(subtype, er), 1) ,3)
```

Which approach do you prefer?

__Challenge 4:__ _Modify the code above to investigate the relationship between Grade and Subtype. Which subtype has the highest proportion of Grade 3 tumours?  Is the association between Grade and Subtupe statistically significant?_

```{r, echo=FALSE, eval=FALSE}
table(subtype, grade)
table(subtype, grade) %>% 
  prop.table(., 1) %>% 
  round(., 3)
```

### Survival analysis

```{r}
plot( survfit(Surv(t.rfs, e.rfs) ~1 ), 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")

plot( survfit(Surv(t.rfs, e.rfs) ~ grade ), col=1:3, 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")
legend('bottomleft', c("G1", "G2", "G3"), fill=1:3)

plot( survfit(Surv(t.rfs, e.rfs) ~ er ), col=1:2, 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")
legend('bottomleft', c("ER-", "ER+"), fill=1:3)
```


```{r]}
plot( survfit(Surv(t.rfs, e.rfs) ~ subtype ), col=1:5, 
      xlab="Time (years)", 
      ylab = "Proportion recurrence free")
groups <- names(table(subtype))
legend('bottomleft', groups, fill=1:5)
```

```{r}
survdiff(Surv(t.rfs, e.rfs) ~ subtype)
```

```{r}
survdiff(Surv(t.rfs, e.rfs) ~ grade)
```

### Gene expression data

Let's find the microarray probe(s) for ESR1:

```{r}
esr1Probes = uppAnnot$probe[ na.omit(uppAnnot$Gene.symbol == 'ESR1') ]
```

```{r, eval=FALSE, echo=FALSE}
par(mfrow=c(2,5))
for(i in esr1Probes){
  x = uppExp[match(i, rownames(uppExp)),]
  boxplot(x ~ er)
}
```

The best one to use is the first one: `esr1Probes[1]`.  What row of the data is this?

```{r}
match("205221_at", rownames(uppExp))
```

```{r}
esr1Dat = uppExp[match("205221_at", rownames(uppExp)), ]
```

To make things easy, lets add this variable to our clinical data:

```{r, warning=FALSE, message=FALSE}
uppClinSmall$esr1Dat = esr1Dat
attach(uppClinSmall)
```

```{r}
ggplot(data=subset(uppClinSmall, !is.na(er)), aes(x=as.factor(er), y=esr1Dat)) + geom_boxplot()
```

### Proliferation genes

The genes below are involved in proliferation:

```{r, echo=FALSE}
#prolifGenes = c("OSTC","MCM6","RPA3","MCM7","PCNA","XRCC6","KPNA2","ANLN","RNASEH2A",
#                "PBK","GMNN","RRM1","CDC45","MAD2L1","RAN","DUT","RRM2","CDK7",
#                "MLH3","SMC4","SMC3","POLD2","POLE2","BCCIP","GINS2","TREX1",
#                "BUB3","FEN1","DBF4B","MOB4","CCNE1","RPA1","POLE3","RFC4","MCM3",
#                "CHEK1","CCND1","CDC37")
prolifGenes = c("MAD2L1", "RRM2", "ANLN", "MCM6", "PBK", "GINS2", "KPNA2", "PCNA")
knitr::kable(data.frame(Gene=prolifGenes, 
                        Probe=uppAnnot$probe[match(prolifGenes, uppAnnot$Gene.symbol)]))
```

```{r}

```

```{r}
## prolifRows = which(uppAnnot$Gene.symbol %in% prolifGenes)
prolifRows = na.omit(match(prolifGenes, uppAnnot$Gene.symbol))
```

```{r}
prolifRows
```

```{r}
prolifDat = uppExp[prolifRows, ]

zz = t(scale(t(prolifDat)))
zz[zz < -3] = -3
zz[zz >  3] =  3

heatmap.2(zz, trace='none', scale='none', col='bluered', 
          labRow = uppAnnot$Gene.symbol[prolifRows])

prolifMean = colMeans(prolifDat)

oo = order(prolifMean)
prolifCol = bluered(length(prolifMean))[rank(prolifMean)]
heatmap.2(zz[,oo], trace='none', scale='none', col='bluered', 
          labRow = uppAnnot$Gene.symbol[prolifRows], Colv=FALSE,
          ColSideColors=prolifCol[oo])


```{r}
ggplot(data=uppClin, aes(x=as.factor(grade), y=prolifMean)) + geom_boxplot()
```

```{r}
ggplot(data=uppClin, aes(x=as.factor(subtype), y=prolifMean)) + geom_boxplot()
```

```{r}
prolifHilo = ifelse(prolifMean > 7, "HighProilif", "LowProlif")

table(prolifHilo)

table(grade, prolifHilo)
```

```{r}
plot( survfit(Surv(t.rfs, e.rfs) ~ prolifHilo ), col=1:2, 
      xlab="Time (years)", ylab = "Proportion recurrence free")
groups <- names(table(prolifHilo))
legend('bottomleft', groups, fill=1:2)
```
